feat: Add IPFS backend

xprazak2 opened the pull request:
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Adds a new backend service provider - IPFS

Xuanwo commented on src/services/ipfs/backend.rs:
```diff
@@ -0,0 +1,162 @@
+#[async_trait]
+impl Accessor for Backend {
+  fn metadata(&self) -> AccessorMetadata {
+    unimplemented!()
```
This API is required to be implemented.

Xuanwo commented on tests/behavior/behavior.rs:
```diff
@@ -147,7 +147,7 @@ macro_rules! behavior_test {
 }
 
-behavior_tests!(Azblob, Fs, Memory, S3);
+behavior_tests!(Azblob, Fs, Memory, S3, ipfs);
```
All items listed here are all `Scheme`. Please add a new variant `Scheme::Ipfs` instead.

Xuanwo commented on src/services/ipfs/tests.rs:
```diff
@@ -0,0 +1,26 @@
+/// In order to test ipfs service, please set the following environment variables:
+/// - `OPENDAL_IPFS_TEST=on`: set to `on` to enable the test.
+/// - `OPENDAL_IPFS_ROOT=/path/to/dir`: set the root dir.
+pub async fn new() -> Result<Option<Arc<dyn Accessor>>> {
```
We have removed all `tests.rs`, please use env to configure the service.

Xuanwo commented on src/services/ipfs/request_writer.rs:
```diff
@@ -0,0 +1,119 @@
+  Sending(BoxFuture<'static, io::Result<()>>),
+}
+
+pub struct IpfsWriterFuture {
```
Please consider use `new_http_channel` in `io_util`.

xprazak2 replied:
I was looking at it previously, but I am not sure how to plug it in. I can use `Box::pin` to pass `fut` into `RequestWriter` and skip `IpfsWriterFuture` but then I get lifetime error:

![io-channel](https://user-images.githubusercontent.com/2664090/184529330-9030f79b-4c95-47c5-a690-cefd1c0336bd.png)

Xuanwo commented on Cargo.toml:
```diff
@@ -75,6 +75,8 @@ reqsign = "0.3.0"
 thiserror = "1.0.30"
 time = "0.3.9"
+ipfs-api-backend-hyper = { version = "0.4", features = ["with-send-sync", "with-builder"] }
+ipfs-api = "0.15"
```
`ipfs-api` has been deprecated, maybe we need to replace it?

Xuanwo commented on src/services/ipfs/backend.rs:
```diff
@@ -0,0 +1,162 @@
+impl Backend {
+    /// Constructor for ipfs builder
+    pub fn new(root: String) -> Self {
+      Self { root, client: IpfsClient::default() }
```
This will only allow connecting the default endpoint `localhost:5001`. Maybe we should allow users to specify the endpoint for the ipfs client?

Maybe using https://docs.rs/ipfs-api-backend-hyper/latest/ipfs_api_backend_hyper/trait.TryFromUri.html#method.from_str ?

Xuanwo commented on src/services/ipfs/backend.rs:
```diff
@@ -0,0 +1,162 @@
+
+impl Backend {
+    /// Constructor for ipfs builder
+    pub fn new(root: String) -> Self {
```
Please expose this API in Builder instead.

Xuanwo reviewed (commented):
Nice PR! I left some reviews, PTAL.

And it's required that all PRs should pass `cargo fmt` and `cargo clippy`, please make sure they are happy.

Xuanwo commented:
@xprazak2 Hi, are you still working on this PR?

xprazak2 commented:
I am still working on it, apologies it is taking me so long. I think I have addressed all the comments except for using `new_http_channel`, I might need a bit more guidance on it. I was looking at it previously, but I am not sure how to plug it in. I can use `Box::pin` to pass fut into `RequestWriter` and skip `IpfsWriterFuture` but then I get lifetime error:
![io-channel](https://user-images.githubusercontent.com/2664090/184530832-e60246e4-52f0-47c5-a752-4b4f1d60eea2.png)

Xuanwo commented on src/services/ipfs/backend.rs:
```diff
@@ -0,0 +1,209 @@
+
+    pub(crate) async fn files_write(&self, path: &str, data: Bytes) -> io::Result<()> {
+        self.client
+            .files_write(path, true, true, data.reader())
```
Sadly, we can't implement `Read` for `Body`. Maybe we need to build a HTTP request directly like we do for `s3` service.

FYI: https://docs.ipfs.tech/reference/kubo/rpc/#api-v0-files-write

BTW: most API calls is so simple that we don't need to use `ipfs-api` SDK.

Xuanwo commented:
We have a refactor in https://github.com/datafuselabs/opendal/pull/556 to make it easier to implement IPFS backend, PTAL.

xprazak2 commented:
Having access to the reader made it much easier to implement `write`. I rebased and removed IPFS SDK.

Xuanwo commented:
> Having access to the reader made it much easier to implement `write`. I rebased and removed IPFS SDK.

Please ping me back when you are ready. And leaving comments for any ideas.

xprazak2 commented:
@Xuanwo, all green now.

Xuanwo commented on src/operator.rs:
```diff
@@ -141,6 +141,7 @@ impl Operator {
                     anyhow!("custom service {v} is not supported"),
                 )))
             }
+            Scheme::Ipfs => services::ipfs::Backend::from_iter(it)?.into(),
```
Move between `Http` and `Memory` to keep the order.

Xuanwo commented on src/services/ipfs/backend.rs:
```diff
@@ -0,0 +1,329 @@
+
+        let req = isahc::Request::post(url.clone());
+        let req = req.body(isahc::AsyncBody::empty()).map_err(|err| {
+            error!("object post: {url} {err:?}");
```
Remove those error logs, as we will print them if `LoggingLayer` is enabled.

Xuanwo commented on src/services/ipfs/backend.rs:
```diff
@@ -0,0 +1,329 @@
+        })?;
+
+        let resp = self.client.send_async(req).await.map_err(|e| {
+            error!("object {path} {op}: {e:?}");
```
ditto

Xuanwo commented on src/services/ipfs/backend.rs:
```diff
@@ -0,0 +1,329 @@
+    async fn make_write_request(
+        &self,
+        query_string: &str,
+        data: &str,
```
We must process input data in an async and streaming way to avoid reading all data into memory.

It's OK for this PR. Please raise a new issue for this so that we can address this in the future.

Xuanwo reviewed (commented):
Mostly LGTM!

Xuanwo commented:
Great, thanks for your PR!
